Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerToContextConnection.cpp

```diff
@@ -467,5 +439,39 @@ void WebSWServerToContextConnection::setInspectable(ServiceWorkerIsInspectable i
     send(Messages::WebSWContextManagerConnection::SetInspectable { inspectable });
 }
 
+void WebSWServerToContextConnection::startProcessAssertionTimer()
+{
+    m_processAssertionTimer.startOneShot(server()->isProcessTerminationDelayEnabled() ? SWServer::defaultTerminationDelay : SWServer::defaultFunctionalEventDuration);
+}
+
+void WebSWServerToContextConnection::serviceWorkerNeedsRunning()
+{
+    if (m_isTakingProcessAssertion)
+        return;
+
+    m_isTakingProcessAssertion = true;
+    protectedConnection()->protectedNetworkProcess()->protectedParentProcessConnection()->send(Messages::NetworkProcessProxy::StartServiceWorkerBackgroundProcessing { webProcessIdentifier() }, 0);
+    startProcessAssertionTimer();
+}
+
+bool WebSWServerToContextConnection::areServiceWorkersIdle() const
+{
+    return m_ongoingFetches.isEmpty() && m_ongoingDownloads.isEmpty() && protectedServer()->areServiceWorkersIdle(*this);
+}
+
+void WebSWServerToContextConnection::processAssertionTimerFired()
+{
+    if (!m_isTakingProcessAssertion)
+        return;
+
+    if (!areServiceWorkersIdle()) {
+        startProcessAssertionTimer();
+        return;
+    }
+
+    m_isTakingProcessAssertion = false;
```

Comment:
What happens if this object in NetworkProcess is destructed before the `EndServiceWorkerBackgroundProcessing` IPC can be sent to the UIProcess?

I tried to somewhat address this issue [286417@main](https://commits.webkit.org/286417@main) in by adding a call to `endServiceWorkerBackgroundProcessing()` in `processHasUnresponseServiceWorker()`, but maybe we need something more generic?

Reply:
In case of network process crash, we terminate all service workers, which calls `disableRemoteWorkers`, which will destroy all process assertions. So we should be good.

There is an edge case we are not covering, which is `WebProcessProxy::m_hasServiceWorkerBackgroundProcessing`.
It could be kept to `true` which could be an issue if the process is again used for service workers.

I'll move WebProcessProxy::m_hasServiceWorkerBackgroundProcessing to WebProcessProxy::m_serviceWorkerInformation::hasBackgroundProcessing.
This way, `disableRemoteWorkers` will reset this flag as well.